### PR TITLE
Add hook to configure EntityEquipmentSlot for item

### PR DESCRIPTION
**Pull request by Landmaster**
This allows mods to create, for example, elytra-like items; this also allows mods to work with `EntityLiving.getSlotForItemStack` correctly.

### CODE REVIEW

**Comment by diesieben07** [comment]
This would allow to deprecate `isValidArmor`.

File: patches/minecraft/net/minecraft/item/Item.java.patch [hunk 1]
```diff
@@ -354,12 +354,14 @@
 +
 +    /**
 +     * Determines if the specific ItemStack can be placed in the specified armor slot.
++     * @deprecated Use {@code armorType == EntityLiving.getSlotForItemStack(stack)} instead.
```
Comment (diesieben07) [inline patches/minecraft/net/minecraft/item/Item.java.patch]: You should probably link to `Item::getEquipmentSlot` here instead, since most uses of this method are overriding it, not calling it.

**Review by diesieben07 — APPROVED** [review]

**Comment by mezz** [comment]
This looks good.
Please update this PR to resolve merge conflicts.

**Comment by Landmaster** [comment]
@mezz Merge conflicts have been resolved.

File: patches/minecraft/net/minecraft/item/Item.java.patch [hunk 2]
```diff
@@ -337,18 +337,34 @@
 +
 +    /**
 +     * Determines if the specific ItemStack can be placed in the specified armor slot.
++     * @deprecated Override {@link #getEquipmentSlot(ItemStack)} to set the armor slot, and use
```
Comment (howtonotwin) [inline patches/minecraft/net/minecraft/item/Item.java.patch]: [`@deprecated` tags go last in the Javadoc](http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#orderoftags). Also, I know you didn't create this method, and you're deprecating it anyway, but can you fix up the rest of the Javadoc while you're at it?

[It should be:](http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#@param)

```
...
@param stack the {@code ItemStack} to check
@param armorType the armor slot to check. ID {@code 0}: helmet; {@code 1}: chest; {@code 2}: legs, {@code 3}: boots		 
@param entity the {@code Entity} trying to equip {@code stack} in the slot specified {@code armorType}
@return whether if {@code stack} can be inserted into the slot
...
```

File: patches/minecraft/net/minecraft/item/Item.java.patch [hunk 3]
```diff
@@ -337,18 +337,34 @@
...
 +     * @return True if the given ItemStack can be inserted in the slot
 +     */
++    @Deprecated
 +    public boolean isValidArmor(ItemStack stack, EntityEquipmentSlot armorType, Entity entity)
 +    {
 +        return net.minecraft.entity.EntityLiving.func_184640_d(stack) == armorType;
 +    }
 +
 +    /**
++     * <em><strong>MODDERS</strong>: Override this to set a non-default {@link EntityEquipmentSlot}
```
Comment (howtonotwin) [inline patches/minecraft/net/minecraft/item/Item.java.patch]: Can you reword this to be less strongly worded? Maybe

```
Override this method to return {@code EntityEquipmentSlot} the given {@code ItemStack} should go in, but <em>do not directly call this to get the {@code EntityEquipmentSlot}</em>. Instead, use {@link ...}.
```

Also, do not link so aggressively. A good rule is to link the first occurrence of a term and then to simply `{@code}` the others. Additionally, try to remove ones that should be known from context even then.

File: patches/minecraft/net/minecraft/item/Item.java.patch [hunk 4]
```diff
@@ -337,18 +337,34 @@
...
++     * for an itemstack, but <strong>do not use this to get the {@link EntityEquipmentSlot} for an
++     * itemstack: for that, use
++     * {@link net.minecraft.entity.EntityLiving#getSlotForItemStack(ItemStack)}.</strong></em>
++     *
++     * @param stack the {@link ItemStack}
++     * @return the {@link EntityEquipmentSlot} for the itemstack, or {@code null} to let
++     * the default vanilla logic as per
++     * {@link net.minecraft.entity.EntityLiving#getSlotForItemStack(ItemStack)} decide
++     */
++    public @Nullable EntityEquipmentSlot getEquipmentSlot(ItemStack stack) { return null; }
```
Comment (howtonotwin) [inline patches/minecraft/net/minecraft/item/Item.java.patch]: No need to squish it down to one line. Expand it out to the whole 4.

**Review by howtonotwin — CHANGES_REQUESTED** [review]
Javadoc nitpicks

File: patches/minecraft/net/minecraft/item/Item.java.patch (hunk 4 above)
Comment (liach) [reply patches/minecraft/net/minecraft/item/Item.java.patch]: Format this code; put annotation in another line in front. You don't need to squeeze the method in one line since it is totally addition, which will not disturb patch application.

File: patches/minecraft/net/minecraft/item/Item.java.patch [hunk 5]
```diff
@@ -352,17 +352,21 @@
 +    }
 +
 +    /**
-+     * <em><strong>MODDERS</strong>: Override this to set a non-default {@link EntityEquipmentSlot}
-+     * for an itemstack, but <strong>do not use this to get the {@link EntityEquipmentSlot} for an
++     * Override this to set a non-default armor slot
++     * for an itemstack, but <em>do not use this to get the armor slot of an
```
Comment (howtonotwin) [inline patches/minecraft/net/minecraft/item/Item.java.patch]: Call it a "stack" or call it `ItemStack` (as code), but don't leave it as "itemstack."

File: patches/minecraft/net/minecraft/item/Item.java.patch [hunk 6]
```diff
@@ -352,17 +352,21 @@
 +    }
 +
 +    /**
-+     * <em><strong>MODDERS</strong>: Override this to set a non-default {@link EntityEquipmentSlot}
-+     * for an itemstack, but <strong>do not use this to get the {@link EntityEquipmentSlot} for an
++     * Override this to set a non-default armor slot
```
Comment (howtonotwin) [inline patches/minecraft/net/minecraft/item/Item.java.patch]: Could you fix the line wrapping?

File: patches/minecraft/net/minecraft/item/Item.java.patch [hunk 7]
```diff
@@ -352,17 +352,21 @@
...
++     * for an itemstack, but <em>do not use this to get the armor slot of an
 +     * itemstack: for that, use
```
Comment (howtonotwin) [inline patches/minecraft/net/minecraft/item/Item.java.patch]: That colon should be a semicolon.

**Review by howtonotwin — CHANGES_REQUESTED** [review]
Final nitpicks.

File: patches/minecraft/net/minecraft/item/Item.java.patch (hunk 5 above)
Comment (liach) [reply patches/minecraft/net/minecraft/item/Item.java.patch]: `item stack`

File: patches/minecraft/net/minecraft/item/Item.java.patch [hunk 8]
```diff
@@ -352,9 +352,8 @@
...
 +
 +    /**
-+     * Override this to set a non-default armor slot
-+     * for an itemstack, but <em>do not use this to get the armor slot of an
-+     * itemstack: for that, use
++     * Override this to set a non-default armor slot for a stack, but <em>do
++     * not use this to get the armor slot of said stack; for that, use
 +     * {@link net.minecraft.entity.EntityLiving#getSlotForItemStack(ItemStack)}.</em>
 +     *
 +     * @param stack the itemstack
```
Comment (howtonotwin) [inline patches/minecraft/net/minecraft/item/Item.java.patch]: I feel really bad saying this, but... you missed a spot. (Also the `@return`.)

Comment (howtonotwin) [reply patches/minecraft/net/minecraft/item/Item.java.patch]: That last commit didn't seem to actually hit the "itemstacks".

File: patches/minecraft/net/minecraft/entity/EntityLiving.java.patch [hunk 9]
```diff
@@ -39,7 +39,15 @@
      public float func_70603_bj()
      {
          return 1.0F;
-@@ -1003,7 +1015,7 @@
+@@ -991,6 +1003,7 @@
+ 
+     public static EntityEquipmentSlot func_184640_d(ItemStack p_184640_0_)
+     {
++        final EntityEquipmentSlot slot = p_184640_0_.func_77973_b().getEquipmentSlot(p_184640_0_); if (slot != null) return slot; // FORGE: Allow modders to set a non-default equipment slot for a stack; e.g. a non-armor chestplate-slot item
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/entity/EntityLiving.java.patch]: Seriously THIS IS TWO LINES OF CODE
Put it on two lines!
Jesus the point of the 'Keep patch files small' is to MAKE THEM READABLE.
To prevent people from reformatting vanilla code because they feel like it. 
So we can see FUNCTIONAL changes vs formatting changes.
If you're just adding a hunk of code, IT DOESN'T need to be all in one line!

File: patches/minecraft/net/minecraft/inventory/ContainerPlayer.java.patch [hunk 10]
```diff
@@ -1,11 +0,0 @@
---- ../src-base/minecraft/net/minecraft/inventory/ContainerPlayer.java
-+++ ../src-work/minecraft/net/minecraft/inventory/ContainerPlayer.java
-@@ -43,7 +43,7 @@
-                 }
-                 public boolean func_75214_a(ItemStack p_75214_1_)
-                 {
--                    return entityequipmentslot == EntityLiving.func_184640_d(p_75214_1_);
-+                    return p_75214_1_.func_77973_b().isValidArmor(p_75214_1_, entityequipmentslot, field_82862_h);
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/inventory/ContainerPlayer.java.patch]: You deprecate this fucntion and do nothing with it to make the new system return the proper value for existing mods with this override.
This is NOT deprecation this is removal. Fix that

**Comment by Landmaster** [comment]
Pinging @LexManos for a second look, now that aforementioned issues have been resolved.

**Comment by LexManos** [comment]
Thanks. This pull request has been integrated manually.
Well thats not descriptive, stupid tool. 
Had to change it because isValidArmor is a different functionality, that lets you have player and slot context.
So things like pumpkins can be equipped, or items can be equipped to multiple slots.
https://github.com/MinecraftForge/MinecraftForge/commit/b4676f5a0bfc3f1049189d247be45bbe53e813f2
